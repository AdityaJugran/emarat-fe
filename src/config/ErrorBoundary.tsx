import React, { Component, ErrorInfo, ReactNode } from 'react';
import * as Sentry from '@sentry/react';

interface Props {
	children: ReactNode;
}

interface State {
	hasError: boolean;
	errorInfo?: ErrorInfo | null;
	eventId?: string;
}

class ErrorBoundary extends Component<Props, State> {
	constructor(props: Props) {
		super(props);

		this.state = {
			hasError: false,
			errorInfo: null,
		};
	}

	public static getDerivedStateFromError(_: Error): State {
		// Update state so the next render will show the fallback UI.
		return { hasError: true };
	}

	public componentDidCatch(error: Error, errorInfo: ErrorInfo) {
		Sentry.withScope((scope) => {
			scope.setExtra('errorInfo', errorInfo);
			const eventId = Sentry.captureException(error);
			this.setState({ eventId, errorInfo });
		});
		this.setState({ errorInfo });
	}

	public render() {
		const { hasError, errorInfo } = this.state;
		if (hasError) {
			return (
				<div className="card my-5">
					<div className="card-header">
						<p>
							There was an error in loading this page.{' '}
							<button
								type="button"
								style={{ cursor: 'pointer', color: '#0077FF' }}
								onClick={() => {
									window.location.reload();
								}}
							>
								Reload this page
							</button>
						</p>
					</div>
					<div className="card-body">
						<details className="error-details">
							<summary>Click for error details</summary>
							{errorInfo?.componentStack.toString()}
						</details>
					</div>
					<button
						type="button"
						className="bg-primary text-light"
						onClick={() =>
							Sentry.showReportDialog({
								eventId: this.state.eventId,
							})
						}
					>
						Report feedback
					</button>
				</div>
			);
		}
		return this.props.children;
	}
}

export default ErrorBoundary;
