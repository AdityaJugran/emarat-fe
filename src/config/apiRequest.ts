import axios from 'axios';
import { toast } from 'react-toastify';
import { API_CONFIG, SERVICE_URL } from 'lib/constants';
import { getAuthToken } from 'lib/firebaseAuth';
import { GenericObject, StringMapObj } from 'lib/types';

type BuildRequestDataType = {
	apiUrl: keyof typeof API_CONFIG;
	headers?: StringMapObj;
	data?: GenericObject;
	appendToUrl?: string;
};

const buildRequestData = async (options: BuildRequestDataType) => {
	const { apiUrl, headers = {}, data, appendToUrl } = options;
	const { url, method } = API_CONFIG[apiUrl];
	const reqUrl = `${SERVICE_URL}${url}${appendToUrl || ''}`;
	const token = await getAuthToken();
	const reqHeaders = {
		'Content-Type': 'application/json',
		Authorizatoin: token ? `Bearer ${token}` : undefined,
		...headers,
	};

	return {
		url: reqUrl,
		headers: reqHeaders,
		data: JSON.stringify(data),
		method,
	};
};

export const apiRequest = async (options: BuildRequestDataType) => {
	const reqData = await buildRequestData(options);
	try {
		const response = await axios(reqData);
		const jsonResponse = await response.data;
		const { meta = {} } = jsonResponse;
		const { success = false, msg = '' } = meta;
		if (success && msg) {
			toast.success(msg);
		}
		return jsonResponse;
	} catch (err) {
		if (err.response) {
			const { data = {} } = err.response;
			const { meta = {} } = data;
			const { msg = '' } = meta;
			toast.error(msg || 'Something Went Wrong');
			return {};
		}
		toast.error('Something went wrong');
	}
	return {};
};
