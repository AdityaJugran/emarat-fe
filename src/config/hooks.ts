import { ROLES } from 'lib/constants';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { useMediaQuery } from 'react-responsive';
import type { RootState, AppDispatch } from 'config/store';
import { setTitle } from 'features/shared/reducers/TopbarSlice';

// Use throughout your app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;

const filterByRole = (isAdmin: boolean) => (option: { role?: string }) => {
	const { role } = option;
	if (!role) return true;
	if (isAdmin && role === ROLES.ADMIN) return true;
	if (!isAdmin && role === ROLES.RESIDENT) return true;
	return false;
};

export const useAuth = () => {
	const authState = useAppSelector((state) => state.auth);
	return { ...authState, filterByRole: filterByRole(authState.isAdmin) };
};

export const useOrientation = () => {
	const isMobile = useMediaQuery({ query: '(max-width: 480px)' });
	return { isMobile };
};


export const useTopbar = () => {
    const { title } = useAppSelector(state => state.topbar);
    const dispatch = useAppDispatch();
    const setUrlTitle = (ti: string) => dispatch(setTitle(ti));
    return { title, setUrlTitle };
}