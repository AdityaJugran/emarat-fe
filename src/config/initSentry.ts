import * as Sentry from '@sentry/react';
import { Integrations } from '@sentry/tracing';
import { SENTRY_DSN } from 'lib/constants';

const initSentry = () => {
	Sentry.init({
		dsn: SENTRY_DSN,
		integrations: [new Integrations.BrowserTracing()],

		// Set tracesSampleRate to 1.0 to capture 100%
		// of transactions for performance monitoring.
		// We recommend adjusting this value in production
		tracesSampleRate: 0.25,
	});
};

export default initSentry;
