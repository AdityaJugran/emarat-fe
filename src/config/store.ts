import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import imageReducer from 'features/shared/components/image/imageSlice';
import authReducer from 'features/shared/reducers/authSlice';
import topbarReducer from 'features/shared/reducers/TopbarSlice';
import socialFeedReducer from 'features/shared/reducers/SocialFeedSlice';

export const store = configureStore({
	reducer: {
		auth: authReducer,
		image: imageReducer,
		topbar: topbarReducer,
		socialFeed: socialFeedReducer,
	},
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
	ReturnType,
	RootState,
	unknown,
	Action<string>
>;
