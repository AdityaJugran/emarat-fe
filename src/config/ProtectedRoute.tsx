import { Redirect, Route, RouteProps } from 'react-router-dom';
import { useAuth } from 'config/hooks';
import { ROLES } from 'lib/constants';

export type ProtectedRouteProps = {
	role?: string;
} & RouteProps;

export default function ProtectedRoute({
	role,
	...routeProps
}: ProtectedRouteProps) {
	const { isLoggedIn, isAdmin } = useAuth();
	const currRole = isAdmin ? ROLES.ADMIN : ROLES.RESIDENT;
	if (isLoggedIn) {
		if (!role) return <Route {...routeProps} />;
		if (role === currRole) return <Route {...routeProps} />;
		return <Redirect to={{ pathname: '/404' }} />;
	}
	return <Redirect to={{ pathname: '/' }} />;
}
