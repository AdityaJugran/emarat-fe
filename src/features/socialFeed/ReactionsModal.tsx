import { useEffect, Dispatch, SetStateAction } from 'react';
import { Modal, Image, Tabs } from 'antd';
import tw from 'twin.macro';
import styled from 'styled-components';
import congratsIcon from 'features/socialFeed/assets/images/congrats.svg';
import laughIcon from 'features/socialFeed/assets/images/laughing.svg';
import likeIcon from 'features/socialFeed/assets/images/like.svg';
import sadIcon from 'features/socialFeed/assets/images/sad.svg';
import userPlaceholderImg from 'assets/images/user-placeholder.svg';
import {
	ReactionType,
	UserReactionListType,
	TotalReactionsType,
} from 'features/socialFeed/SocialFeedTypes';
import { useSocialFeed } from 'config/hooks';
import { loadUsers } from 'features/socialFeed/firebase/reactions';

type PropsType = {
	isVisible: boolean;
	setIsVisible: Dispatch<SetStateAction<boolean>>;
	postId: string;
	postOwner: string;
	reactions?: UserReactionListType;
	totalReactions: TotalReactionsType;
};

export default function ReactionTabs({
	isVisible,
	setIsVisible,
	postId,
	postOwner,
	reactions,
	totalReactions,
}: PropsType) {
	const { users, addUser } = useSocialFeed();
	const title = postOwner !== '' ? `Reactions on ${postOwner}'s post` : '';

	useEffect(() => {
		loadUsers(postId, users, addUser);
	}, []);

	const reactionUI = (type?: ReactionType) =>
		reactions ? (
			Object.entries(reactions)
				.reverse()
				.filter(
					([key, reaction]) =>
						type === undefined || reaction.reaction === type
				)
				.map(([key, reaction]) => {
					const {
						firstName = '',
						lastName = '',
						picture,
					} = users[key] || {};

					return (
						<div key={key} className="flex items-center py-2">
							<Image
								className="rounded-full"
								height={35}
								width={35}
								preview={false}
								src={picture || userPlaceholderImg}
								fallback={userPlaceholderImg}
								alt="User Avatar"
							/>
							<p className="ml-3">{`${firstName} ${lastName}`}</p>
						</div>
					);
				})
		) : (
			<span />
		);

	return (
		<Modal
			title={<h3 className="text-xl">{title}</h3>}
			visible={isVisible}
			footer={null}
			onCancel={() => setIsVisible(false)}
			destroyOnClose={true}
			centered
		>
			<Tabs defaultActiveKey="1">
				<Tabs.TabPane tab={`All  ${totalReactions.sum}`} key="1">
					{reactionUI()}
				</Tabs.TabPane>
				{totalReactions.like > 0 && (
					<Tabs.TabPane
						tab={
							<ReactionContainer>
								<Reaction src={likeIcon} alt="Like Icon" />
								{totalReactions.like}
							</ReactionContainer>
						}
						key="2"
					>
						{reactionUI('like')}
					</Tabs.TabPane>
				)}

				{totalReactions.laugh > 0 && (
					<Tabs.TabPane
						tab={
							<ReactionContainer>
								<Reaction src={laughIcon} alt="Laugh Icon" />
								{totalReactions.laugh}
							</ReactionContainer>
						}
						key="3"
					>
						{reactionUI('laugh')}
					</Tabs.TabPane>
				)}

				{totalReactions.sad > 0 && (
					<Tabs.TabPane
						tab={
							<ReactionContainer>
								<Reaction src={sadIcon} alt="Sad Icon" />
								{totalReactions.sad}
							</ReactionContainer>
						}
						key="4"
					>
						{reactionUI('sad')}
					</Tabs.TabPane>
				)}
				{totalReactions.congrats > 0 && (
					<Tabs.TabPane
						tab={
							<ReactionContainer>
								<Reaction
									src={congratsIcon}
									alt="Congrats Icon"
								/>
								{totalReactions.congrats}
							</ReactionContainer>
						}
						key="5"
					>
						{reactionUI('congrats')}
					</Tabs.TabPane>
				)}
			</Tabs>
		</Modal>
	);
}

const Reaction = styled.img`
	${tw`
		w-10
		px-2
	`}
`;

const ReactionContainer = styled.div`
	${tw`
		flex
		items-center
	`}
`;
