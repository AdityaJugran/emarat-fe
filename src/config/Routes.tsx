import { Redirect, Route, Switch } from 'react-router-dom';
import { lazy, Suspense } from 'react';
import ProtectedRoute from 'config/ProtectedRoute';
import ManageUsers from 'features/users/ManageUsers';
import Home from 'features/home/Home';
import ManageAmenities from 'features/amenities/ManageAmenities';
import MyAmenities from 'features/amenities/MyAmenities';
// import MyPayments from 'features/payments/MyPayments';
// import Dashboard from 'features/dashboard/Dashboard';
// import SocialFeed from 'features/socialFeed/SocialFeed';
import ManageComplaints from 'features/complaints/ManageComplaints';
import MyComplaints from 'features/complaints/MyComplaints';
import Transactions from 'features/payments/Transactions';
import Broadcasts from 'features/broadcasts/Broadcasts';
import PageNotFound from 'features/errorPages/404';
import { ROLES } from 'lib/constants';
import Profile from 'features/profile/Profile';
import { useAuth } from 'config/hooks';
import Announcements from 'features/announcements/Announcements';

const MyPayments = lazy(() => import('features/payments/MyPayments'));
const Dashboard = lazy(() => import('features/dashboard/Dashboard'));
const SocialFeed = lazy(() => import('features/socialFeed/SocialFeed'));

export default function Routes() {
	const { isAdmin } = useAuth();

	return (
		<Suspense fallback={<div>loading...</div>}>
		<Switch>
			<Route path="/" exact>
				<Home />
			</Route>
			
			<Route path="/404" exact>
				<PageNotFound />
			</Route>
			<ProtectedRoute
				path="/users"
				role={ROLES.ADMIN}
				component={ManageUsers}
				exact
			/>
			<ProtectedRoute
				path="/amenities"
				component={isAdmin ? ManageAmenities : MyAmenities}
				exact
			/>
			
			<ProtectedRoute
				path="/transactions"
				role={ROLES.ADMIN}
				component={Transactions}
				exact
			/>
			
			<ProtectedRoute
				path="/dashboard"
				role={ROLES.ADMIN}
				component={Dashboard}
				exact
			/>
			<ProtectedRoute 
			path="/social-feed"
			component={
				SocialFeed
			}
			exact
			/>
			<ProtectedRoute
				path="/payments"
				role={ROLES.RESIDENT}
				component={MyPayments}
				exact
			/>
			
			<ProtectedRoute
				path="/complaints"
				role=""
				component={isAdmin ? ManageComplaints : MyComplaints}
				exact
			/>
			<ProtectedRoute
				path="/announcements"
				role={ROLES.RESIDENT}
				component={Announcements}
				exact
			/>
			
			<ProtectedRoute path="/profile" component={Profile} exact />
			<ProtectedRoute path="/broadcasts" component={Broadcasts} exact />
			<Redirect from="*" to="/404" />
			
		</Switch>
		</Suspense>
	);
}
