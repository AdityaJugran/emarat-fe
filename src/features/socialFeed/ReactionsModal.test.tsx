import ReactionsModal from 'features/socialFeed/ReactionsModal';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { store } from 'config/store';

it('Reactions Modal', () => {
	const tree = renderer
		.create(
			<Provider store={store}>
				<ReactionsModal
					isVisible={false}
					setIsVisible={jest.fn()}
					postId={postIdMock}
					postOwner={postOwnerMock}
					totalReactions={totalReactionsMock}
				/>
			</Provider>
		)
		.toJSON();
	expect(tree).toMatchSnapshot();
});

const postIdMock = '-MjP6ufjLnHm8rWFc_wx';

const postOwnerMock = 'John';

const totalReactionsMock = {
	like: 10,
	laugh: 0,
	sad: 10,
	congrats: 40,
	sum: 60,
};
