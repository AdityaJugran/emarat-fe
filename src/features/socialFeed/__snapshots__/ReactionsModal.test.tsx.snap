// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reactions Modal 1`] = `null`;
